lightning/src/ln/channelmanager.rs

```diff
@@ -1528,29 +1530,71 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 	/// Note that ALL inputs in the transaction pointed to by funding_txo MUST spend SegWit outputs
 	/// or your counterparty can steal your funds!
 	///
+	/// Returns an [`APIError::APIMisuseError`] if the funding_transaction spent non-SegWit outputs
+	/// or if no output was found which matches the parameters in [`Event::FundingGenerationReady`].
+	///
 	/// Panics if a funding transaction has already been provided for this channel.
 	///
-	/// May panic if the funding_txo is duplicative with some other channel (note that this should
-	/// be trivially prevented by using unique funding transaction keys per-channel).
-	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_txo: OutPoint) {
+	/// May panic if the output found in the funding transaction is duplicative with some other
+	/// channel (note that this should be trivially prevented by using unique funding transaction
+	/// keys per-channel).
+	///
+	/// Do NOT broadcast the funding transaction yourself. When we have safely received our
+	/// counterparty's signature the funding transaction will automatically be broadcast via the
+	/// [`BroadcasterInterface`] provided when this `ChannelManager` was constructed.
+	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_transaction: Transaction) -> Result<(), APIError> {
 		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
 
+		for inp in funding_transaction.input.iter() {
+			if inp.witness.is_empty() {
+				return Err(APIError::APIMisuseError {
+					err: "Funding transaction must be fully signed and spend Segwit outputs".to_owned()
+				});
+			}
+		}
+
 		let (chan, msg) = {
 			let (res, chan) = match self.channel_state.lock().unwrap().by_id.remove(temporary_channel_id) {
 				Some(mut chan) => {
-					(chan.get_outbound_funding_created(funding_txo, &self.logger)
+					let mut output_index = None;
+					let expected_spk = chan.get_funding_redeemscript().to_v0_p2wsh();
+					for (idx, outp) in funding_transaction.output.iter().enumerate() {
+						if outp.script_pubkey == expected_spk && outp.value == chan.get_value_satoshis() {
+							if output_index.is_some() {
+								return Err(APIError::APIMisuseError {
+									err: "Multiple outputs matched the expected script and value".to_owned()
+								});
+							}
+							if idx > u16::max_value() as usize {
```

Comment:
If you assume minimal-sized output (8 bytes value + 2 bytes scriptPubkey), given that MAX_STANDARD_TX_WEIGHT is about 100k bytes, I think you can just reject a transaction with an idx above 10_000 ?

Reply:
Sure, but its nice to get an error instead of a panic, even if you did something kinda stupid.